aten/src/ATen/native/TensorCompare.cpp

```diff
@@ -239,6 +241,128 @@ Tensor wrapped_scalar_tensor(
 
+Tensor& isin_out(const Tensor& elements, const Tensor& test_elements, bool assume_unique, bool invert, Tensor& out) {
+  out.resize_as_(elements);
+  if (elements.numel() == 0) {
+    return out;
+  }
+
+  // Condition taken from numpy's implementation.
+  Tensor test_elements_flat = test_elements.ravel();
+  if (test_elements_flat.numel() < static_cast<int64_t>(
+        10.0f * std::pow(static_cast<double>(elements.numel()), 0.145))) {
+    auto iter = TensorIteratorConfig()
+      .set_check_mem_overlap(false)
+      .add_output(out)
+      .add_input(elements)
+      .cast_common_dtype_to_outputs(false)
+      .enforce_safe_casting_to_output(false)
+      .check_all_same_dtype(false)
+      .build();
+    isin_default_stub(iter.device_type(), iter, test_elements_flat, invert);
+    return out;
+  }
+
+  // If the number of test elements is large, use a sorting-based algorithm:
+  // 1. Concatenate unique elements with unique test elements in 1D form. If
+  //    assume_unique is true, skip calls to unique().
+  // 2. Stable sort all elements, maintaining order indices to reverse the
+  //    operation. Stable sort is necessary to keep elements before test
+  //    elements within the sorted list.
+  // 3. Create a mask for locations of adjacent duplicate values within the
+  //    sorted list. Duplicate values are in both elements and test elements.
+  // 4. Reorder the mask to match the pre-sorted element order.
+  // 5. Index the mask to match the pre-unique element order. If
+  //    assume_unique is true, just take the first N items of the mask,
+  //    where N is the original number of elements.
+
+  Tensor elements_flat, unique_order;
+  if (assume_unique) {
+    elements_flat = elements.ravel();
+  } else {
+    std::tie (elements_flat, unique_order) = at::_unique(
+        elements, /*sorted=*/ false, /*return_inverse=*/ true);
+    std::tie (test_elements_flat, std::ignore) = at::_unique(test_elements_flat, /*sorted=*/ false);
+  }
+
+  Tensor all_elements = at::_cat({elements_flat, test_elements_flat});
+  Tensor sorted_elements, sorted_order;
+  std::tie (sorted_elements, sorted_order) = all_elements.sort(
+      /*stable=*/ c10::optional<bool>(true), /*dim=*/ 0, /*descending=*/ false);
+
+  Tensor duplicate_mask = at::empty(all_elements.sizes(),
+    TensorOptions(elements.device()).dtype(ScalarType::Bool));
+  auto iter = TensorIteratorConfig()
+    .set_check_mem_overlap(false)
+    .add_output(duplicate_mask)
+    .add_input(sorted_elements)
+    .cast_common_dtype_to_outputs(false)
+    .enforce_safe_casting_to_output(false)
+    .check_all_same_dtype(false)
+    .build();
+  AT_DISPATCH_INTEGRAL_TYPES(elements.scalar_type(), "isin_sorting", [&]() {
+    iter.for_each([&](char** data, const int64_t* strides, int64_t n) {
+      auto* mask_bytes = data[0];
+      const auto* element_bytes = data[1];
+
+      for (int i = 0; i < n - 1; ++i) {
+        auto* mask_data = reinterpret_cast<bool*>(mask_bytes);
```

Comment:
nit: Do we really need `reinterpret_cast` here? I mean for all the places that using `reinterpret_cast` within this function.

Reply:
Yes, because the data pointers are `char*` and the compiler doesn't allow `static_cast`s from `char*` to `scalar_t*` or `bool*`. I think if the data pointers were `void*` instead, `static_cast` would work.

Reply:
Also, if the stride math is done on a `scalar_t*`, it will be wrong, so the cast has to be done every iteration of the loop.

Reply:
Ah, I see. I was thinking that you can use: 
`bool* mask_data = (bool *)mask_bytes;`
Let's use reinterpret_cast then.